File: backend/__SLUG__/src/middleware/request_log.go
package middleware

import (
	"net/http"
	"time"

	"github.com/felixge/httpsnoop"
	"go.uber.org/zap"
	"go_backend/src/logger"
)

// RequestLog logs method, path, status, and duration for every request.
func RequestLog(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		metrics := httpsnoop.CaptureMetrics(next, w, r)
		logger.LoggerFromContext(r.Context()).Debug("request complete",
			zap.String("method", r.Method),
			zap.String("path", r.URL.Path),
			zap.Int("status", metrics.Code),
			zap.Int64("duration_ms", time.Since(start).Milliseconds()),
		)
	})
}

File: backend/__SLUG___node/src/middleware/requestLog.ts
import type { NextFunction, Request, Response } from "express";
import { loggerFromContext } from "../logger/logger";

/** Logs method, path, status, and duration for every request. */
export function requestLog(req: Request, res: Response, next: NextFunction): void {
  const start = Date.now();
  res.on("finish", () => {
    loggerFromContext().debug(
      {
        method: req.method,
        path: req.path,
        status: res.statusCode,
        duration_ms: Date.now() - start,
      },
      "request complete",
    );
  });
  next();
}
